### PR TITLE
feat: Complex slice inputs for dynamic slice builtins

### LINKED ISSUE #3364: Enable using complex type inputs on dynamic slice intrinsics
### Problem

We can appropriately use a dynamic slice intrinsic on a nested slice whose element types are primitive types, but we should be able to also use dynamic slice intrinsics on a nested slice whose element types are also complex types such as another nested slice.

### Happy Case

If I have this nested slice:
```
struct FooParent {
    parent_arr: [Field; 3],
    foos: [Foo],
}

struct Bar {
    inner: [Field; 3],
}

struct Foo {
    a: Field,
    b: [Field],
    bar: Bar,
}

    let q = x.push_back(foo_four);
    let foo_parent_one = FooParent { parent_arr: [0, 1, 2], foos: x };
    let foo_parent_two = FooParent { parent_arr: [3, 4, 5], foos: q };
    let mut foo_parents = [foo_parent_one];
    foo_parents = foo_parents.push_back(foo_parent_two);
```
I should be able to do:
```
    assert(foo_parents[y - 2].foos.len() == 5);
    foo_parents[y - 2].foos = foo_parents[y - 2].foos.push_back(foo_four);
    assert(foo_parents[y - 2].foos.len() == 6);
```
the same way I can do:
```
    foo_parents[y - 2].foos[y - 1].b = foo_parents[y - 2].foos[y - 1].b.push_back(500);
    assert(foo_parents[y - 2].foos[y - 1].b.len() == 7);
    assert(foo_parents[y - 2].foos[y - 1].b[6] == 500);
```
where the nested slice simply has primitive element types. 

### Alternatives Considered

N/A

### Additional Context

Nested dynamic slices were made possible by this pass (https://github.com/noir-lang/noir/pull/3258). There is a chance that something is simply off in this pass rather than the ACIR gen. 

After some investigation it does look like the fill internal slices pass is working as expected but this is still important to note in case I missed something. 

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #3364 

## Summary\*

This PR heavily changes the way slice intrinsics are implemented in ACIR gen. Following the `fill_internal_slices` pass nested slices now have dummy data automatically attached to them to enable known sizes to be used when fetching from an array with nested slices. Most of the slice intrinsics were written for non-complex inputs and also did not account for the addition of already included dummy data. With dummy data already included we cannot simply push back or push front elements for example, we need to make sure we do not alter the size of the internal slice we are using. 

The slice intrinsic implementations essentially treat the input already as a dynamic array and perform the appropriate memory operations upon them. I considered alternatives and went with what I think is the most efficient way to implement these intrinsics, however, there is definitely room for cleanup and in the name of getting a first iteration working, we can consider how to optimize these intrinsics in another PR. 

I have added comments in the ACIR gen itself to help with some of the logic for the implementation, but do let me know if something is not clear and I can document it further. 

The `fill_internal_slices` pass now also accounts for slice constants passed as inputs to slice intrinsics as these must also be filled with dummy data when they are being attached to a nested slice.

## Additional Context

A bug was found in the `fill_internal_slices` pass that was necessary to resolve this issue. Essentially the children fetched from a slice were not being accurately tracked when an `array_get` occurs. This fix is included in this PR as it did not come up until more complex operations with nested slices were attempted. 

SliceInsert and SliceRemove now use a dynamic index. I only tested using a constant index in this PR and will have a followup PR that shows this tested out as this PR is already large enough.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [X] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — APPROVED** [review]
I'm still definitely worried the slice handling code is too complex for one feature. This PR looks good but I could be missing something since it is hard to grok the whole context.

**Comment by vezenovm** [comment]
> I'm still definitely worried the slice handling code is too complex for one feature. This PR looks good but I could be missing something since it is hard to grok the whole context.

Thanks for the review. Yeah slice handling has gotten quite complex with the introduction of arbitrarily nested slices. As @TomAFrench is out today I was going to let him take a look before merging. He has been looking at how to optimize some of the dynamic array code and although I think a followup PR is better to add optimizations, perhaps there are some obvious inefficiencies I missed.

**Review by TomAFrench — COMMENTED** [review]
I've made a few suggestions for readability so that the smoothbrains like me can follow slice logic more easily.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs [hunk 1]
```diff
@@ -1771,29 +1771,56 @@ impl Context {
                     self.check_array_is_initialized(arguments[1], dfg)?;
                 let slice = self.convert_value(arguments[1], dfg);
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:1773]: Yeah good call. I will unpack through the other intrinsics as well.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs [hunk 2]
```diff
@@ -1807,155 +1834,474 @@ impl Context {
...
+
+                    let same_value_read_var =
+                        self.acir_context.read_from_memory(block_id, &index_var)?;
+
+                    let not_pred = self.acir_context.sub_var(one, greater_eq_than_idx)?;
+                    let false_pred = self.acir_context.mul_var(not_pred, same_value_read_var)?;
+
+                    let new_value = self.acir_context.add_var(true_pred, false_pred)?;
+
+                    self.acir_context.write_to_memory(result_block_id, &index_var, &new_value)?;
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:2291]: I agree I'll apply some renames

[the remark this answers is not in the document]
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs]: Pushed some renames and a few additional comments. Let me know if that is any clearer.

File: compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs [hunk 3]
```diff
@@ -1807,155 +1834,474 @@ impl Context {
...
+                        popped_elements.push(element);
+                    }
+                }
+
+                // Go through the entire slice argument and determine what value should be written to the new slice.
+                // 1. If the current index is greater than the removal index we must write the next value
+                //    from the original slice to the current index
+                // 2. At the end of the slice reading from the next value of the original slice
+                //    can lead to a potential out of bounds error. In this case we just read from the original slice
+                //    at the current index. As we are decreasing the slice in length, this is a safe operation.
```
[the remark this answers is not in the document]
Comment (vezenovm) [reply compiler/noirc_evaluator/src/ssa/acir_gen/mod.rs:2273]: Yeah good call. I changed this, let me know if what I did is what you were thinking. Doing so also showed me a place where I was doing an unnecessary clone on initialization. 

I also added a sanity check that the flattened `Vector<AcirValue>` represented by `new_slice` matches the computed flattened size.

**Review by TomAFrench — APPROVED** [review]
Just some nits and a potential optimisation.

The optimisation isn't a blocker as we can always follow up with another PR.

**Pull request merged**